activesupport/lib/active_support/tagged_logging.rb

```diff
@@ -29,56 +29,86 @@ module TaggedLogging
     module Formatter # :nodoc:
       # This method is invoked when a log event occurs.
       def call(severity, timestamp, progname, msg)
-        if t = tags_text
-          super(severity, timestamp, progname, t.concat(msg.to_s))
-        else
-          super(severity, timestamp, progname, msg)
-        end
+        super(severity, timestamp, progname, tag_stack.format_message(msg))
       end
 
       def tagged(*tags)
-        new_tags = push_tags(*tags)
+        pushed_count = tag_stack.push_tags(tags).size
         yield self
       ensure
-        pop_tags(new_tags.size)
+        pop_tags(pushed_count)
       end
 
       def push_tags(*tags)
-        tags.flatten!
-        tags.reject!(&:blank?)
-        current_tags.concat tags
-        tags
+        tag_stack.push_tags(tags)
       end
 
-      def pop_tags(size = 1)
-        current_tags.pop size
+      def pop_tags(count = 1)
+        tag_stack.pop_tags(count)
       end
 
       def clear_tags!
-        current_tags.clear
+        tag_stack.clear
       end
 
-      def current_tags
+      def tag_stack
         # We use our object ID here to avoid conflicting with other instances
-        thread_key = @thread_key ||= "activesupport_tagged_logging_tags:#{object_id}"
-        IsolatedExecutionState[thread_key] ||= []
+        @thread_key ||= "activesupport_tagged_logging_tags:#{object_id}"
+        IsolatedExecutionState[@thread_key] ||= TagStack.new
+      end
+
+      def current_tags
+        tag_stack.tags
       end
 
       def tags_text
-        tags = current_tags
-        if tags.one?
-          +"[#{tags[0]}] "
-        elsif tags.any?
-          tags.collect { |tag| "[#{tag}] " }.join
+        tag_stack.format_message("")
+      end
+    end
+
+    class TagStack # :nodoc:
+      attr_reader :tags
+
+      def initialize
+        @tags = []
+        @tags_string = nil
+      end
+
+      def push_tags(tags)
+        @tags_string = nil
+        tags.flatten!
+        tags.reject!(&:blank?)
+        @tags.concat(tags)
+        tags
+      end
+
+      def pop_tags(count)
+        @tags_string = nil
+        @tags.pop(count)
+      end
+
+      def clear
+        @tags_string = nil
+        @tags.clear
+      end
+
+      def format_message(message)
+        if @tags.empty?
+          message
+        elsif @tags.size == 1
+          "[#{@tags[0]}] #{message}"
+        else
+          @tags_string ||= "[#{@tags.join("] [")}] "
```

Comment:
Wouldn't it make sense to add the surounding `[]` when the tag is pushed so that in the end you only need to `join(" ")`?

Reply:
Do you mean like:

```ruby
class TagStack
  def push_tags(tags)
    @tags_string = nil
    tags.flatten!
    tags.reject!(&:blank?)
    tags.map! { |tag| "[#{tag}]" } # this?
    @tags.concat(tags)
    tags
  end
end
```

That would change the return value of `current_tags`, and would break, for example:

https://github.com/rails/rails/blob/26e115e58e91a8b0f1012f33f46b6d26a0a90de9/activesupport/lib/active_support/tagged_logging.rb#L137

Also, it would negate the O(N) => O(1) allocations optimization:

```
Warming up --------------------------------------
messages: 1, tags: 0    31.490k i/100ms
messages: 1, tags: 1     8.716k i/100ms
messages: 1, tags: 2     8.778k i/100ms
messages: 1, tags: 3     8.023k i/100ms
messages: 3, tags: 1     4.743k i/100ms
messages: 3, tags: 2     5.227k i/100ms
messages: 3, tags: 3     4.813k i/100ms
Calculating -------------------------------------
messages: 1, tags: 0    314.550k (± 1.6%) i/s -      1.574M in   5.006885s
messages: 1, tags: 1     88.169k (± 1.5%) i/s -    444.516k in   5.042753s
messages: 1, tags: 2     86.736k (± 1.5%) i/s -    438.900k in   5.061355s
messages: 1, tags: 3     79.837k (± 1.3%) i/s -    401.150k in   5.025438s
messages: 3, tags: 1     47.818k (± 1.5%) i/s -    241.893k in   5.059867s
messages: 3, tags: 2     51.758k (± 1.2%) i/s -    261.350k in   5.050220s
messages: 3, tags: 3     48.097k (± 1.5%) i/s -    240.650k in   5.004513s
Calculating -------------------------------------
messages: 1, tags: 0   166.000  memsize (     0.000  retained)
                         2.000  objects (     0.000  retained)
                         1.000  strings (     0.000  retained)
messages: 1, tags: 1   562.000  memsize (     0.000  retained)
                         8.000  objects (     0.000  retained)
                         3.000  strings (     0.000  retained)
messages: 1, tags: 2   486.000  memsize (     0.000  retained)
                         9.000  objects (     0.000  retained)
                         4.000  strings (     0.000  retained)
messages: 1, tags: 3   726.000  memsize (     0.000  retained)
                        10.000  objects (     0.000  retained)
                         4.000  strings (     0.000  retained)
messages: 3, tags: 1     1.366k memsize (     0.000  retained)
                        16.000  objects (     0.000  retained)
                         3.000  strings (     0.000  retained)
messages: 3, tags: 2   978.000  memsize (     0.000  retained)
                        15.000  objects (     0.000  retained)
                         4.000  strings (     0.000  retained)
messages: 3, tags: 3     1.538k memsize (     0.000  retained)
                        16.000  objects (     0.000  retained)
                         4.000  strings (     0.000  retained)
```

Reply:
I mean in another array, but yeah you are right, it means more strings allocs, probably not worth it 👍